### PR TITLE
[SPARK-31734][ML][PySpark] Add weight support in ClusteringEvaluator

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add weight support in ClusteringEvaluator

### Why are the changes needed?
Currently, BinaryClassificationEvaluator, RegressionEvaluator, and MulticlassClassificationEvaluator support instance weight, but ClusteringEvaluator doesn't, so we will add instance weight support in ClusteringEvaluator.

### Does this PR introduce _any_ user-facing change?
Yes.
ClusteringEvaluator.setWeightCol

### How was this patch tested?
add new unit test

### CODE REVIEW

**Review by zhengruifeng — APPROVED** [review]
LGTM

**Comment by huaxingao** [comment]
cc @srowen

**Review by srowen — COMMENTED** [review]
I may have missed it but does this check somewhere that the weights are positive?

**Comment by huaxingao** [comment]
Added the check to make sure weights are all positive.
We probably need to do this for everything that has instance weight support. I will open a ticket to fix all of them?

**Comment by srowen** [comment]
Yeah there are several points in the code that check weights with a `require`. I don't know if they can be refactored easily. Your approach is fine. Other places check it farther downstream inside some other operation but it doesn't matter much.

**Comment by huaxingao** [comment]
I mean in the following algorithms that have instance weight support, we probably also need to make sure the weights are positive.

- DecisionTreeClassifier/Regressor
- GBTClassifier/Regressor
- LinearSVC
- LogisticRegression
- RandomForestClassifier/Regressor
- LinearRegression
- GeneralizedLinearRegression
- KMean
- BisectiongKMean
- GaussianMixture
- BinaryClassificationEvaluation
- MulticlassClassificationEvaluation
- RegressionEvaluation

I will take a look and fix them all. Not in this PR, though.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 1]
```diff
@@ -360,43 +379,47 @@ private[evaluation] object SquaredEuclideanSilhouette extends Silhouette {
...
         seqOp = {
           case (
-              (featureSum: DenseVector, squaredNormSum: Double, numOfPoints: Long),
-              (features, squaredNorm)
+              (featureSum: DenseVector, squaredNormSum: Double, weightSum: Double),
+              (features, squaredNorm, weight)
             ) =>
-            BLAS.axpy(1.0, features, featureSum)
-            (featureSum, squaredNormSum + squaredNorm, numOfPoints + 1)
+            require (weight >= 0.0, "illegal weight value: " + weight + " weight must be >= 0.0")
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala:404]: I think it's better to do the check here so it doesn't require an extra pass to get all the weights.

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala:404]: Looks good; it's consistent with your other change. Really minor: use string interpolation?

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 2]
```diff
@@ -555,30 +583,35 @@ private[evaluation] object CosineSilhouette extends Silhouette {
...
-      .aggregateByKey[(DenseVector, Long)]((Vectors.zeros(numFeatures).toDense, 0L))(
+      .map { row => (row.getDouble(0), (row.getAs[Vector](1), row.getDouble(2))) }
+      .aggregateByKey[(DenseVector, Double)]((Vectors.zeros(numFeatures).toDense, 0.0))(
       seqOp = {
-        case ((normalizedFeaturesSum: DenseVector, numOfPoints: Long), (normalizedFeatures)) =>
-          BLAS.axpy(1.0, normalizedFeatures, normalizedFeaturesSum)
-          (normalizedFeaturesSum, numOfPoints + 1)
+        case ((normalizedFeaturesSum: DenseVector, weightSum: Double),
+          (normalizedFeatures, weight)) =>
+          require (weight >= 0.0, "illegal weight value: " + weight + " weight must be >= 0.0")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala:607]: Same here and nit: remove space after require

**Comment by huaxingao** [comment]
I didn't resolve conflicts correctly. I will fix the problem.

**Comment by srowen** [comment]
Merged to master. You may want to further change the nonnegativity check in your other PR to use the new method you introduced there.

**Comment by huaxingao** [comment]
Thanks! @srowen @zhengruifeng 
I have updated the nonnegativity check in the other PR.

**Pull request closed without merging**